stage0/src/asm/boot.s

```diff
@@ -167,33 +142,24 @@ _protected_mode_start:
     xor %ecx, %ecx            # ECX = 0 - we're not interested in a subpage
     cpuid                     # EAX, EBX, ECX, EDX = CPUID(EAX, ECX)
     and $0b111111, %ebx       # zero out all but EBX[5:0], which the C-bit location
-    push %ebx                 # save the full C-bit location value for later
+    mov %ebx, %esi            # save the full C-bit location value for later to pass into main
     sub $32, %ebx             # let's assume the encrypted bit is > 32, as it simplifies logic below
     mov $1, %edi
     mov %ebx, %ecx
-    shl %cl, %edi             # construct the bit mask for the loop, store it in EDI
-
-    # Note that this sets the encrypted bit for _all_ entries in the page tables, even
-    # if they are unused. This is fine, as those entries will still not have the PRESENT
-    # bit set and thus will be ignored by the CPU.
-    mov ${pml4}, %eax         # Base values for all page tables.
-    mov ${pdpt}, %ebx
-    mov ${pd_0}, %ecx
-    mov ${pd_3}, %edx
-    mov ${pt_0}, %ebp
-    xor %esi, %esi            # zero out esi for use as the loop counter
-    1:
-    # Page table entries are 8 bytes and the encrypted bit is in the second 4 bytes,
-    # so these offsets boil down to base + (ESI * 8) + 4.
-    or %edi, 4(%eax,%esi,8)   # PML4[ESI] |= EDI
-    or %edi, 4(%ebx,%esi,8)   # PDPT[ESI] |= EDI
-    or %edi, 4(%ecx,%esi,8)   # PD_0[ESI] |= EDI
-    or %edi, 4(%edx,%esi,8)   # PD_3[ESI] |= EDI
-    or %edi, 4(%ebp,%esi,8)   # PT_0[ESI] |= EDI
-    inc %esi                  # ESI = ESI + 1
-    cmp $512, %esi            # have we changed all 512 entries?
-    jne 1b                    # ESI < 512, go back to square 1
-    pop %esi                  # keep track of the encrypted bit for main
+    shl %cl, %edi             # construct the encrypted bit mask, store it in EDI
+
+    # We set the encrypted bit for each of the page table entries that we previously created.
+    # The encrypted bit is in the second half of each 8-byte entry, so we add an extra offset of 4 bytes.
+    mov ${pml4}, %eax
```

Comment:
I think you can collapse these down to just one `mov`, just like with the first half.